### ISSUE TITLE
model: publish an accepted RecurrentPPO checkpoint with immutable Hugging Face provenance

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7851]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Require accepted #7849 checkpoint, normalizer/action/recurrent descriptors, durable evidence and checkpoint-specific rights; #7850 affects only optional system context.

### Ordered implementation / investigation plan

1. Require accepted #7849 checkpoint, normalizer/action/recurrent descriptors, durable evidence and checkpoint-specific rights; #7850 affects only optional system context.
2. Assemble model card, exact configs/manifests, rights/notices and SHA256SUMS without uploading or committing weight bytes.
3. After exact publication approval, upload to the selected namespace and record immutable HF revision; do not rely on mutable latest/tag.
4. Cold-download into an empty cache, verify all hashes and stateful inference parity, then merge the registry pointer with inherited evidence tier.

### Files and owners to inspect

- model/registry.yaml
- docs/policy_cards/
- model registry resolver/tests
- compact publication receipt

### Acceptance and validation

- Raw policy and policy-plus-wrapper results remain separate.
- No missing normalizer, unknown rights or unverified upload enters the registry.

### Current dependencies and issue-text corrections

- Dependencies: #7849 #6855 or exact rights equivalent; publication approval
- Keep publication blocked until exact checkpoint selection and rights resolve; remove any language implying namespace suggestion is authorization.
- Record exact external revision, cold-verification receipt and supported observation/action/recurrent contract.
- The supplied draft was checked against the refreshed live issue and complete comment thread; historical rationale is retained below.
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:blocked`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `dependency:has-blockers`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Parent and Dependencies

- Programme: #7845
- Minimum empirical prerequisite: #7849 records an eligible accepted checkpoint and bounded evidence tier.
- Wrapper-system context: #7850, when executed, determines whether a wrapper configuration should be linked as a separate system component.
- Rights prerequisite: #6855 or an equivalent exact checkpoint-specific rights disposition.
- Position: child F, publication and registry integration only.

## Archetype Metadata

```yaml
archetype: model-publication
evidence_tier: inherits-reviewed-source-evidence
resource: network-artifact-publication
publication_authorized: false
```

## Objective

Publish one empirically accepted, rights-cleared RecurrentPPO checkpoint to an immutable Hugging Face model repository and bind it into Robot SF's existing model registry with complete observation, action, recurrent-state, training, evaluation, checksum, license, and limitation metadata.

Publication must make the exact model reproducible and retrievable. It must not strengthen the evidence tier, imply deployment safety, or publish checkpoint bytes whose redistribution basis is unresolved.

## Publication Boundary

The published object is the **raw RecurrentPPO policy checkpoint** selected under #7849. The planner-agnostic safety wrapper is code/configuration, not part of the learned checkpoint.

When #7850 exists, the model card may link an exact reviewed wrapper configuration as an optional system composition. It must distinguish:

```text
raw RecurrentPPO policy
RecurrentPPO + named wrapper configuration
```

Do not label wrapper-on outcomes as raw-policy performance.

## Preconditions

Before any upload or public repository creation:

1. #7849 identifies one eligible checkpoint through the preregistered selection rule.
2. Checkpoint bytes, normalizer/state assets, config, source, and evaluation packet have verified SHA-256 digests.
3. The checkpoint's observation and action contracts resolve exactly.
4. Training and evaluation manifests have durable retrieval pointers.
5. #6855 or a checkpoint-specific legal/rights record states:
   - copyright holder;
   - redistribution basis;
   - weights origin;
   - training-code license;
   - training-data/scenario provenance and license status;
   - required notices;
   - allowed and blocked distribution modes.
6. The intended Hugging Face namespace, repository visibility, repository license metadata, and retention policy are explicitly selected.
7. Domain/model-publication review approves the exact bytes and metadata packet.

Missing rights or provenance remains blocked. Do not infer a weights license from the Robot SF source-code license.

## Proposed Repository

Suggested name, subject to namespace and rights confirmation:

```text
ll7/robot-sf-recurrent-ppo-local-planner-v1
```

Use one dedicated model repository rather than placing unrelated policy families in a shared unversioned location.

The exact Hugging Face commit revision after upload is the immutable external identity. Tags or branches are convenience pointers only and cannot replace the revision SHA.

## Required Artifact Contents

At minimum:

```text
README.md
model.zip
training_config.yaml
resolved_config.yaml
training_manifest.json
evaluation_manifest.json
evaluation_results.json
source_and_environment.json
observation_contract.json
action_contract.json
recurrent_state_contract.json
checkpoint_index.json
scenario_split_manifest.json
limitations.json
LICENSE or rights statement
THIRD_PARTY_NOTICES.md where required
PROVENANCE.json
SHA256SUMS
```

Include a normalizer or preprocessing asset only when required and bind its checksum. If the policy needs a normalizer that cannot be produced or verified, publication remains blocked.

Do not upload:

- credentials or private storage paths;
- raw private training data;
- unredacted scheduler/host details that are not needed for reproducibility;
- unrelated checkpoints;
- mutable `latest` artifacts without immutable revision binding;
- local-only or checksum-unverified outputs.

## Model Card Contract

The Hugging Face `README.md` must contain structured metadata and readable sections covering:

### Model identity

- policy ID and version;
- algorithm: `sb3_contrib.RecurrentPPO`;
- policy class: `MultiInputLstmPolicy`;
- Hugging Face immutable revision;
- Robot SF source commit;
- checkpoint SHA-256 and size;
- config and dependency-lock digests;
- training seed;
- selected-checkpoint rule and source evidence.

### Intended use

- Robot SF simulator local-planning research;
- exact supported observation/action/kinematics contract;
- deterministic benchmark inference through the stateful adapter from #7848;
- named supported Robot SF versions or source revisions.

### Not intended for

- direct physical-vehicle deployment;
- real-world pedestrian-intention inference;
- universal planner ranking;
- safety certification or safety guarantee;
- use with different observation shapes, timestep, action units, robot dynamics, or normalizers without revalidation;
- treating wrapper-on evidence as raw-model evidence.

### Observation contract

State exact:

- Gymnasium space type;
- keys `drive_state` and `rays`;
- shapes, dtypes, bounds, coordinate frames, units, normalization, and missing-data behavior;
- prohibited privileged/future fields;
- observation contract digest.

### Action contract

State exact:

- raw model output representation and bounds;
- desired unicycle velocity units `(m/s, rad/s)`;
- canonical conversion to linear/angular acceleration using `(target-current)/dt`;
- supported `dt`, speed, acceleration, deceleration, and angular limits;
- action contract digest;
- requirement to use the Robot SF adapter rather than sending model output directly to hardware.

### Recurrent-state contract

State exact:

- LSTM hidden size and layers;
- shared/separate actor/critic LSTM settings;
- state tensor structure and expected batch/environment dimension;
- `episode_start` behavior;
- required reset events;
- prohibition on cross-episode, cross-worker, or cross-robot state reuse;
- deterministic inference setting.

### Training

- training scenario manifest and digest;
- training seed and environment-step budget;
- reward and termination contract;
- optimizer/hyperparameters;
- software and hardware environment;
- evaluation cadence;
- checkpoint-selection rule;
- interruptions/resume history;
- known training anomalies or unavailable fields.

### Evaluation

- exact evaluation scenario/seed manifests;
- evidence tier and claim boundary;
- episode/scenario counts and missingness;
- collision-free completion, success, collision, near-miss/exposure, timeout/low-progress, clearance, efficiency, comfort, and latency results supported by #7849;
- training-seed variability and uncertainty;
- raw versus wrapper-on results as separate tables when #7850 applies;
- failure/degraded/fallback exclusions.

### Limitations and risks

- simulator and pedestrian-model dependence;
- perfect or simplified sensing assumptions where applicable;
- limited scenario and kinematics scope;
- possible hidden-state distribution shift;
- sensitivity to timestep, action conversion, observation normalization, and checkpoint/source version;
- collision and liveness failure modes;
- absence of sim-to-real evidence.

## Robot SF Registry Integration

Add or update one exact row in `model/registry.yaml` and generated documentation according to current policy. The row must contain or reference:

- stable model ID;
- Hugging Face repository and immutable revision;
- file path within the repository;
- SHA-256 and size;
- source/training commit;
- training and evaluation config paths/digests;
- observation/action/recurrent contract identifiers;
- normalizer URI/checksum or explicit `not_required`;
- policy/algorithm class;
- intended planner adapter;
- supported kinematics and `dt`;
- evidence and benchmark status;
- license/rights fields;
- known limitations and blocked uses.

Model resolution must download or fetch the exact immutable revision and verify the checkpoint checksum before load. No mutable branch or tag may be accepted without the expected digest.

## Publication Procedure

1. Build the publication directory from verified source artifacts in a clean task-owned workspace.
2. Validate every required file and cross-reference.
3. Generate deterministic `SHA256SUMS` over the published file set.
4. Run model-card and registry-schema checks.
5. Create or update the Hugging Face repository only after rights and review approval.
6. Upload the exact packet.
7. Record the resulting immutable Hugging Face revision.
8. Fetch/download the packet into a fresh clean directory by immutable revision.
9. Reverify every checksum, file count, model load, observation/action contract, and deterministic inference smoke.
10. Only after cold verification, merge the Robot SF registry pointer and documentation.

If the Hugging Face upload helper does not natively support SB3-Contrib RecurrentPPO, use the generic repository/file API. Do not coerce the checkpoint into a different algorithm class for helper compatibility.

## Validation and Cold Verification

Required checks include:

- local checkpoint SHA-256 versus publication `SHA256SUMS`;
- Hugging Face downloaded bytes versus local bytes;
- immutable revision resolution;
- Robot SF registry resolution and checksum enforcement;
- load through the #7848 adapter;
- observation/action/recurrent contract validation;
- deterministic fixed-observation-sequence inference parity before and after upload;
- no fallback/degraded model resolution;
- model-card required-section and metadata validation;
- rights/notice file presence and consistency.

One suggested cold smoke:

```text
fresh checkout + empty model cache
  -> resolve registry model at immutable HF revision
  -> verify SHA-256
  -> construct #7848 adapter
  -> replay fixed observation sequence
  -> compare commands and recurrent reset diagnostics with source fixture
```

## Monitoring and Operational Safety

During upload and verification, record:

- actor and tool version;
- source file inventory and total bytes;
- upload start/finish status;
- returned repository and revision identity;
- retry/failure class;
- downloaded file inventory and total bytes;
- checksum and load verdicts;
- registry resolution status.

Do not print tokens, authorization headers, private paths, or raw authenticated responses. An interrupted or partial upload remains unpublished/unverified and cannot be referenced by the registry.

## Acceptance Criteria

- [ ] One exact #7849 checkpoint is selected through the preregistered rule and has durable source evidence.
- [ ] Checkpoint-specific rights and redistribution status are explicit and approved.
- [ ] All required model-card, contract, provenance, rights, result, and checksum files are complete.
- [ ] Hugging Face stores the exact packet at one immutable revision.
- [ ] A fresh cold download matches every published checksum and file identity.
- [ ] The model loads through #7848 and reproduces the deterministic inference fixture.
- [ ] `model/registry.yaml` resolves the immutable revision and fails closed on missing or changed bytes.
- [ ] Raw-model and optional wrapper-system claims remain separate.
- [ ] No checkpoint bytes are committed to ordinary Git history.
- [ ] Publication does not promote the evidence tier or modify dissertation/paper claims automatically.
- [ ] The final repository, immutable revision, checksum, rights record, and cold-verification receipt are posted to #7845.

## Suggested Paths

- `model/registry.yaml`
- `model/registry.md` or its generated owner
- `docs/policy_cards/`
- model resolution and validation tests
- one compact publication/cold-verification evidence directory

Keep generated checkpoint bytes outside Git. Use Hugging Face and the approved durable source store for binary custody.

## Stop Conditions

- Stop on unknown or conflicting checkpoint rights.
- Stop if the selected model, normalizer, config, source, or evidence digest cannot be reproduced.
- Stop if the model card would need to omit a required observation/action/recurrent limitation.
- Stop if cold download or deterministic parity fails.
- Stop if registry resolution would depend only on a mutable branch/tag.
- Stop before publication if #7849 is invalid/inconclusive or does not admit a checkpoint.

## Claim Boundary

Completion proves immutable publication, retrieval, provenance, and adapter compatibility for one named checkpoint. It does not establish deployment readiness, safety certification, real-world validity, universal superiority, or a stronger empirical result than the source evidence from #7849/#7850.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 7851
source_body_sha256: f809a69995451c21f0cb616cc6708e260c928c84bff31ff248c109662d0cfdbb
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: blocked
next_action: resolve_named_blocker
authority: blocker_owner
execution_mode: blocker
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-04
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
